tests/bsim/bluetooth/host/att/timeout/main.c

```diff
@@ -0,0 +1,275 @@
+
+#define UUID_2                                                                                     \
+	BT_UUID_DECLARE_128(0x3f, 0xa4, 0x7f, 0x44, 0x2e, 0x2a, 0x43, 0x05, 0xab, 0x38, 0x07,      \
+			    0x8d, 0x16, 0xbf, 0x99, 0xf1)
+
+static bool trigger_att_timeout;
+static K_SEM_DEFINE(disconnected_sem, 0, 1);
+
+static ssize_t read_chrc(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			void *buf, uint16_t buf_len, uint16_t offset)
+{
+	ssize_t read_len;
+
+	LOG_INF("ATT timeout will %sbe triggered", trigger_att_timeout ? "" : "not ");
+
+	if (trigger_att_timeout) {
+		/* Sleep longer than ATT Timeout (section 3.3.3). */
+		k_sleep(K_SECONDS(ATT_TIMEOUT_SEC + 1));
+	}
+
+	__ASSERT_NO_MSG(offset == 0);
+	read_len = buf_len;
+
+	__ASSERT_NO_MSG(read_len >= 2);
+	sys_put_le16(read_len, buf);
+
+	return read_len;
+}
+
+static struct bt_gatt_attr attrs[] = {
+	BT_GATT_PRIMARY_SERVICE(UUID_1),
+	BT_GATT_CHARACTERISTIC(UUID_2, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_chrc, NULL, NULL),
+};
+
+static struct bt_gatt_service svc = {
+	.attrs = attrs,
+	.attr_count = ARRAY_SIZE(attrs),
+};
+
+static void bs_sync_all_log(char *log_msg)
+{
+	/* Everyone meets here. */
+	bt_testlib_bs_sync_all();
+
+	if (get_device_nbr() == 0) {
+		LOG_WRN("Sync point: %s", log_msg);
+	}
+
+	/* Everyone waits for d0 to finish logging. */
+	bt_testlib_bs_sync_all();
+}
+
+static inline void bt_enable_quiet(void)
+{
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_ERR);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_ERR);
+
+	EXPECT_ZERO(bt_enable(NULL));
+
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_INF);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_INF);
+}
+
+/**
+ * Test procedure:
+ *
+ * Central:
+ * 1. Connect to the peripheral.
+ * 2. Try to read a characteristic value.
+ * 3. Expect BT_ATT_ERR_UNLIKELY error.
+ * 4. Expect the peripheral to disconnect.
+ * 5. Reconnect to the peripheral.
+ * 6. Try to read a characteristic value.
+ * 7. Expect the peripheral to respond with the characteristic value.
+ * 8. Ensure that connection stays alive after a delay equal to ATT timeout.
+ * 9. Disconnect from the peripheral.
+ *
+ * Peripheral:
+ * 1. Start advertising.
+ * 2. Make the read callback sleep for more than ATT Timeout when the central tries to read.
+ * 3. Expect the disconnected callback to be called.
+ * 4. Start advertising again.
+ * 5. Make the read callback respond with the characteristic value when the central tries to read.
+ * 6. Expect the connection stay alive after a delay equal to ATT timeout.
+ * 7. Expect the central to disconnect.
+ */
+static void test_timeout(enum bt_att_chan_opt bearer)
+{
+	bool central = (get_device_nbr() == CENTRAL_DEVICE_NBR);
+	bool peripheral = (get_device_nbr() == PERIPHERAL_DEVICE_NBR);
+	bt_addr_le_t adva;
+	struct bt_conn *conn = NULL;
+	int err;
+	uint16_t handle = 0;
+
+	/* Two iterations:
+	 * 1. ATT timeout.
+	 * 2. Successful read.
+	 */
+	for (int i = 0; i < 2; i++) {
```

Comment:
Hmm, this loop looks a bit weird to me. 

It contains both `if(peripheral)` and `if(central)` stuff, as well as `if i == 0` and `if i == 1`.

Have you considered defining the behavior as functions? e.g. something along the lines of

```c
test_timeout_peripheral(...)
{
    test_peripheral_timeout();
    test_peripheral_read();
}

test_timeout_central
{
    test_central_timeout();
    test_central_read();
}

static void test_timeout(enum bt_att_chan_opt bearer)
{
	bool central = (get_device_nbr() == CENTRAL_DEVICE_NBR);
	bool peripheral = (get_device_nbr() == PERIPHERAL_DEVICE_NBR);

	if (central) {
		test_timeout_central();
	} else if (peripheral) {
		test_timeout_peripheral();
	}

}
```

Which gives a better overview than a long loop with several switch cases and very little common code.

Reply:
Tried to simplify a bit. But kept the loop. This way it is easier to see what happens on each side at each test step as well as when sync point is called. Ultimately, this is a test, not the final application.

Reply:
I agree with Emil. This loop is a bad abstraction. The `if (timeout) { bt_testlib_conn_unref(&conn); }` is really spaghetti. It's relying on `timeout` being true only in the last iteration of the loop. With this kind of logic, you might as well put the `if (central) {bt_testlib_disconnect}` in the else clause. Why not unroll the loop?

Reply:
Tests are supposed to be simple. Otherwise we will need tests for the tests. :P

Reply:
Done